experimental/Schemes/Schemes.jl

```diff
@@ -0,0 +1,2587 @@
+export Covering
+export patches, glueings, _add_patch!
+
+export Refinement
+
+export CoveredScheme
+export coverings
+
+export CovSchMorphism
+
+export AbstractCoherentSheaf
+
+export IdealSheaf
+export parent, covering, ideals
+
+export CoherentSheaf
+export parent, covering, modules
+
+export LineBundle, VectorBundle
+export parent, covering, transitions, rank, OO, direct_sum
+
+export VectorBundleSection, LineBundleSection
+export tautological_sections
+
+export projectivize
+export cartesian_product
+
+@Markdown.doc """
+    abstract type Scheme{ S <: Ring }
+
+The type parameter S is the type of the base ring k over which the Scheme is defined.
+"""
+abstract type Scheme{ S <: Ring }end
+@doc Markdown.doc"""
+    abstract type AffineScheme{S, T <: MPolyRing, U <: MPolyElem} <: Scheme{S}
+
+The type parameters are: 
+ * T: The type of the ambient polynomial ring R
+ * U: The type of the elements in R
+"""
+abstract type AffineScheme{S, T <: MPolyRing, U <: MPolyElem} <: Scheme{S} end
+abstract type SchemeMorphism end
+
+abstract type SchematicPoint end
+abstract type Sheaf end
+abstract type AbstractCoherentSheaf <: Sheaf end
+
+@doc Markdown.doc"""
+    mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
+
+Models an affine scheme X = Spec R/I with R = k[x₁,…,xₙ] a free 
+polynomial algebra of type T over a base ring k of type S and I ⊂ R a finitely generated ideal 
+with elements of type U.
+"""
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
+  # the basic fields 
+  k::S			# the base ring (usually a field) of definition for the scheme
+  R::T		  	# the ambient polynomial ring to model this affine scheme
+  I::MPolyIdeal{U}	# The ideal in R defining the scheme
+
+  # fields for caching
+  name::String # the name of this scheme for printing
+
+  function Spec(k::S, R::T, I::MPolyIdeal{U} ) where{
+			S <: Ring, T <:MPolyRing , U <: MPolyElem}
+    if k != coefficient_ring(R)
+      error("Base ring of the affine scheme does not coincide with the base ring of the associated algebra")
+    end
+    # TODO: Implement further plausibility checks to be performed at runtime.
+    return new{S, T, U}(k, R, I )
+  end
+end
+
+# The copy constructors.
+# In this case, it's easy, since non of the used data structures 
+# can be modified from the outside with "Fernwirkung" affecting 
+# our instance of Spec. 
+#
+# To explain what I mean by "Fernwirkung": Say, your struct stores 
+# a vector v as a member variable and v is passed on to the constructor. 
+# Simply storing v is wrong, because it is an object that could potentially
+# also be accessed from other parts of the program. If v is altered, then 
+# also the member variable of our Spec is altered! Thus in this case, 
+# we would really need to copy v before storing it in the Spec.
+#
+# Such behaviour is not expected for instances of rings or ideals. 
+# As far as I understand, these objects are supposed to behave like 
+# static ones to the user. But we should watch out for exceptions and 
```

Comment:
What do you mean with "static" here? Immutable perhaps?

(Nb, there is a slight tension between "immutable" in the Julia sense, which I'd call "read-only", which is not recursive; and "immutable" in a more mathematical sense (that's how GAP uses it) which is kinda recursive and means "this will be the same mathematical object, nothing is gonna change that, although the object itself may change (i.e. may not be read-only) to store extra data (such as computed properties, additional charts, whatever)